### PR TITLE
Tutorial: NumPy NLP from scratch with a focus on ethics

### PR DESCRIPTION
**Pull request by Dbhasin1**
Hi @melissawm @isabela-pf @rgommers 
> This tutorial demonstrates how to build a simple <a href = 'https://en.wikipedia.org/wiki/Long_short-term_memory'> Long Short Term memory network (LSTM) </a> from scratch in NumPy to perform sentiment analysis on a socially relevant and ethically acquired dataset.
> ## Table of contents
>   - Data Collection 
>   - Preprocess the datasets
>   - Build and train a LSTM network from scratch
>   - Perform sentiment analysis on collected speeches 
>   - Next steps

I have also added a file called [text_preprocessing.py](https://github.com/Dbhasin1/numpy-tutorials/blob/ethics-tutorial/content/text_preprocessing.py) which includes basic text preprocessing functions which would have made the tutorial too lengthy had I included them there. 

Also, I modified `requirements.txt` and `environement.yml` to include `pandas` however I'm not sure if I should include `re` and `string` which I use for text preprocessing because I believe they are Python-Built-in Modules. 

Any feedback will be most welcome, Thank you!!!

### CODE REVIEW

**Comment by melissawm** [comment]
(We are aware of the CI issues and will try to figure those out, so no need to review that part of it now)

**Review by rossbar — COMMENTED** [review]
As mentioned above, the big blocker here is getting the data incorporated. I held off on taking a deeper look at the code itself, but I'll happily take another pass once we figure out how to surmount the "getting data and making this executable" hurdle.

The big thing that jumped out to me was the text preprocessing. Rather than have this in a separate file, I'd actually prefer it to be incorporated in the notebook itself. The main advantage of doing this is that the code will be run with each build, which reduces the chance that the preprocessing code gets out of sync with the notebook itself. It also makes it easier for the user to modify if they're working interactively and want to evaluate what effect the preprocessing has on the results. I definitely understand the desire not to distract from the narrative with a giant code block. I think a nice solution would be to add the code to a cell that is [toggled hidden](https://myst-nb.readthedocs.io/en/latest/use/hiding.html#hiding-code-cells). 

I also went ahead and pushed up a couple minor changes to a) add the tutorial to the site and b) get rid of some sphinx warnings.

**Comment by melissawm** [comment]
Thanks @rossbar ! This is something that we can certainly do - you're right, our biggest concern was this being a distraction from the rest of the tutorial (and making it much longer) but if it can be toggled hidden I think that is a nice compromise.

**Comment by Dbhasin1** [comment]
Thank you @rossbar, I agree with @melissawm and will incorporate the toggle in my next commit. 
Also, with regards to caching the data, the few options we could find are:

-  Using the [Intake](https://github.com/intake/intake) library to host the dataset on a remote server, which we can direct the readers to, or submit a package to conda, which can include the data files themselves, and people can download the dataset using `conda install ...`
-  For the `glove` embeddings, we could store the dictionary in a JSON file and add that to the repo, we could also convert the IMDB dataframe to NumPy arrays and store them in a `.npy` file, but I will have to modify a major chunk of `text_preprocessing` in that case, as it is mainly intended to work for dataframes 

These are mainly what we've thought of so far. Let me know what you think!

**Comment by rgommers** [comment]
A few thoughts on the dataset storage issue: Intake seems nicer than a conda packages. Intake is meant for exactly this type of use case, and a conda package would be a bit of a hack.

For storing dataframe data it'd be good to store it in a modern high-performance binary format (unless it's quite small and JSON is fine) - `parquet` is the first one that comes to mind.

**Comment by rgommers** [comment]
I just read through the tutorial text: it is well-structured and reads quite well, looks like excellent work @Dbhasin1! I need to find some time to actually play with it, but based on the demo you showed earlier it looks like this is shaping up very nicely.

> Also, I modified `requirements.txt` and `environement.yml` to include `pandas` however I'm not sure if I should include `re` and `string` which I use for text preprocessing because I believe they are Python-Built-in Modules.

Indeed, `re` and `string` are in the Python standard library, so no need to add them.

File: content/tutorial-nlp-from-scratch.md [hunk 1]
```diff
@@ -0,0 +1,848 @@
...
++++
+
+You have learned how to build and train a simple Long Short Term Memory network from scratch using just NumPy to perform sentiment analysis.
+
+To further enhance and optimize your neural network model, you can consider one of a mixture of the following:
+
+- Increase the training sample size by increasing the `split_percentile`.
+- Alter the architecture by introducing multiple LSTM layers to make the network deeper.
+- Use a higher epoch size to train longer and add more regularization techniques, such as early stopping, to prevent overfitting.
+- Introduce a validation set for an unbiased valuation of the model fit.
```
Comment (rgommers) [inline content/tutorial-nlp-from-scratch.md:835]: "valuation" -> "evaluation"

File: content/tutorial-nlp-from-scratch.md [hunk 2]
```diff
@@ -0,0 +1,848 @@
...
+- Apply batch normalization for faster and more stable training.
+- Tune other parameters, such as the learning rate and hidden layer size.
+- Initialise weights using [Xavier Initialisation](https://d2l.ai/chapter_multilayer-perceptrons/numerical-stability-and-init.html) to prevent vanishing/exploding gradients instead of initialising them randomly.
```
Comment (rgommers) [inline content/tutorial-nlp-from-scratch.md:838]: Maybe you wanted to link directly to https://d2l.ai/chapter_multilayer-perceptrons/numerical-stability-and-init.html#xavier-initialization?

**Review by rgommers — COMMENTED** [review]
I've now read it in detail (not grabbed all the data yet and ran it). Still liking the overall structure, and the explanations are very clear. All my comments are quite minor at this point.

File: content/tutorial-nlp-from-scratch.md [hunk 3]
```diff
@@ -0,0 +1,848 @@
...
+4. Perform sentiment analysis on collected speeches 
+
+5. Next steps
+
++++
+
+## 1. Data Collection
+
+Before you begin there are a few pointers you should always keep in mind before choosing the data you wish to train your model on:
+- **Identifying Data Bias** - Bias is a component of the human thought process, and data collected from humans therefore inherently reflects that bias. Some ways in which this bias tends to occur in Machine Learning datasets are:
```
[the remark this answers is not in the document]
Comment (Dbhasin1) [reply content/tutorial-nlp-from-scratch.md:58]: it was intended to be `from` because we're basically trying to convey that data sourced from human activities can reflect the bias in their minds. I'll try rephrasing it

File: content/tutorial-nlp-from-scratch.md [hunk 4]
```diff
@@ -0,0 +1,848 @@
...
+```{code-cell} ipython3
+# Importing the necessary packages 
+import numpy as np 
+import pandas as pd 
+import matplotlib.pyplot as plt 
+from text_preprocessing import TextPreprocess 
+import string
+```
+
+````{admonition}**Define a class to perform the aforementioned text preprocessing and conversion techniques on the input data**
```
[the remark this answers is not in the document]
Comment (Dbhasin1) [reply content/tutorial-nlp-from-scratch.md:125]: I asked @melissawm about this, it should render when the `.md` file is converted to `.html`

File: content/tutorial-nlp-from-scratch.md [hunk 5]
```diff
@@ -0,0 +1,848 @@
...
+
+### Introduction to a Long Short Term Memory Network
+
+In an artificial neural network (ANN), the information only moves in one direction — from the input layer, through the hidden layers, to the output layer. The information moves straight through the network and never takes the previous nodes into account at a later stage. Because it only considers the current input, the features learned are not shared across different positions of the sequence. Moreover, it cannot process sequences with varying lengths.
+
+Unlike an ANN, the RNN was designed to work with sequence prediction problems.RNNs introduce state variables to store past information, together with the current inputs, to determine the current outputs. Since an RNN shares the learned features with all the data points in a sequence regardless of its length, it is capable of processing sequences with varying lengths.
```
[the remark this answers is not in the document]
Comment (Dbhasin1) [reply content/tutorial-nlp-from-scratch.md:365]: Yep, that makes sense. How about I replace it with `Multilayer perceptron (MLP)`?

[the remark this answers is not in the document]
Comment (rgommers) [reply content/tutorial-nlp-from-scratch.md:365]: That should work.

File: content/tutorial-nlp-from-scratch.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (melissawm) [reply content/tutorial-nlp-from-scratch.md:125]: I think we talked about the toggle not this one. If we want the notebook to be executable I think we should stick to regular markdown syntax.

[the remark this answers is not in the document]
Comment (melissawm) [reply content/tutorial-nlp-from-scratch.md:125]: Just to clarify: if we use the admonition directive, this works fine for the html generated pages, but not for the executable notebook. The [Myst-NB](https://myst-nb.readthedocs.io/en/latest/use/hiding.html) documentation highlights the option of using a cell tag to toggle hiding cells from the html input. This works fine except that in the executable notebook the cell is not hidden. I think this is ok, since if people to download the executable notebook they probably want to see the entire code. So I'd suggest the following: 
- Remove the admonition and `:class: toggle` options, marking this cell as a regular executable python/code cell;
- Edit the cell metadata to add the `hide-input` tag. You can do this in jupyter-lab by clicking the Property Inspector on the right toolbar and adding a tag to the selected cell:
Screenshot_20210923_163408

[the remark this answers is not in the document]
Comment (Dbhasin1) [reply content/tutorial-nlp-from-scratch.md:125]: Understood, Thanks @melissawm!

**Comment by Dbhasin1** [comment]
Hello, I just made a few commits incorporating the changes @rgommers suggested along with tidying up the code a bit. 
We're also experimenting with a library called [pooch](https://www.fatiando.org/pooch/latest/index.html) which basically downloads the requisite datasets from the URLs we hardcode. It creates a cache and stores all the data over there, so hopefully, our files won't be downloaded each time the CI builds. We're using pooch to download the IMDb dataset and the glove embeddings only. As for the `speeches.csv` file, we just manually upload it to the repository as there is no pre-existing URL for that particular file that we can provide pooch with.  Apologies for how long it took to deal with the data issue. Hopefully, the notebook will be completely reproducible now

**Comment by rgommers** [comment]
Nice, thanks @Dbhasin1 

> We're also experimenting with a library called [pooch](https://www.fatiando.org/pooch/latest/index.html) which basically downloads the requisite datasets from the URLs we hardcode. It creates a cache and stores all the data over there, so hopefully, our files won't be downloaded each time the CI builds.

This sounds like a very reasonable solution. It would perhaps be useful to still ping the original URL to ensure it's available. Otherwise a problem may only surface once the CI cache expires. Not a huge issue perhaps, not sure. Maybe pooch has functionality for this already?

**Comment by Dbhasin1** [comment]
> This sounds like a very reasonable solution. It would perhaps be useful to still ping the original URL to ensure it's available. Otherwise a problem may only surface once the CI cache expires. Not a huge issue perhaps, not sure. Maybe pooch has functionality for this already?

pooch uses what are called `hashes` to verify that the file being downloaded is the same as what we'd originally intended it to be. I'm working right now on obtaining the hashes and allocating them to the respective URLs in the registry. In case, the URL is not available in the future I think it will throw a normal 404 error, and in case the URL is available however the file at the same location has been modified it will throw an error that the hashes of both the files do not match. More about this can be read [here](https://www.fatiando.org/pooch/latest/hashes.html). But in case the URL turns out to be unavailable, do we incorporate any alternate solution to obtain the data in the tutorial?

**Comment by rgommers** [comment]
> But in case the URL turns out to be unavailable, do we incorporate any alternate solution to obtain the data in the tutorial?

I think it's a matter of estimating how reliable external links are. If it's the canonical link for a well-known dataset, the problem is very likely to be temporary so we just wait. If the link is to something less well-known, we can keep a backup somewhere or anyway mirror the data and change the link.

**Comment by Dbhasin1** [comment]
Maybe we can also try maintaining our own repository on [zenodo](https://zenodo.org/) and then fetching the files from those URLs via pooch? Zenodo's by CERN so it guarantees data longevity as long as the user does not manually withdraw their data from the platform. 
>  we can keep a backup somewhere or anyway mirror the data and change the link.

I'm not sure how the latter is done, I'll look into it once

File: content/tutorial-nlp-from-scratch.ipynb [hunk 6]
```diff
@@ -0,0 +1,1458 @@
...
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "fba4c540",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "def backprop(y, caches, hidden_dim, input_dim, time_steps, parameters):\n",
+    "    # Retrieve output and corresponding weights of fully connected layer\n",
```
Comment (czgdp1807) [inline content/tutorial-nlp-from-scratch.ipynb:854]: May be this can be broken down into different functions according to the gates present in LSTM. For example, first a forget gate is present which takes current word in the sequence (x_t) and the hidden state of previous cell (h_t_prev). A function can be added,
```python
def forget_gate(x_t, h_t_prev):
    # do stuff
   return forget
```
There are lots of `x`, `+`, sigma, `tanh` which have their own role in the LSTM model, so writing functions with meaningful names for each of them will be great. These can be then called at appropriately places inside `backprop` and forward pass function (don't know the name of the function implemented for this). Let me know if you want to discuss this further.

"Step by Step LSTM Walk Through" section in https://colah.github.io/posts/2015-08-Understanding-LSTMs/ can help in modularising this thing. Cheers.

Comment (Dbhasin1) [reply content/tutorial-nlp-from-scratch.ipynb]: Hi @czgdp1807, apologies for the late response. I really liked your suggestions and have incorporated the suggested changes in the latest commit along with the addition of a small diagram that gives an overview of the gate mechanisms happening inside a memory block of the LSTM. Let me know what you think!

Comment (czgdp1807) [reply content/tutorial-nlp-from-scratch.ipynb]: Looks good. It's easier to understand now.

**Comment by rgommers** [comment]
I'm happy with that.

**Comment by melissawm** [comment]
Folks, this has been sitting for a while so I'll merge and any further enhancements can be done in a follow-up. Thanks again @Dbhasin1 !

**Comment by Dbhasin1** [comment]
Thank you everyone for the immense help in getting the tutorial published! 
I observed that the toggle on the text preprocessing failed to render, along with a few images and a gif. Is there something we can do to correct this?

**Comment by melissawm** [comment]
Ah - good catch, I completely forgot to check. Here's the situation:

- We can't use html tags for images because of the way the jupyterbook is built. When building through jupyterbook, images get put into a separate `_images` folder which makes the html tag (which goes raw into the final html doc) point to the wrong place.
- For the toggle, I tried every possible combination but couldn't make it work. Cell tags don't seem to be considered at all when rendering. Not sure if I'm misunderstanding something... [Here's the relevant documentation](https://myst-nb.readthedocs.io/en/latest/use/hiding.html).

#114 is supposed to fix the image display problem.
